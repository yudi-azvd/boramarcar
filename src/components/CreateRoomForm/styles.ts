import styled from "styled-components";

// Ferra a responsividade do Ant Design
export const InputGroup = styled.div`
  /* display: flex;
  justify-content: space-between;

  .ant-form-item {
    width: 48%;
  } */
`