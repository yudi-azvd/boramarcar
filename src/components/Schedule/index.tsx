import React, { useState } from "react"
import { Container, Timebox } from "./style"

export type Day = 'Sunday' | 'Monday' | 'Tuesday' | 'Wednesday' | 'Thursday' | 'Friday' | 'Saturday'

export type TimeboxValue = 'available' | 'busy' | undefined

export interface ScheduleProps {
  data: {
    times: string[], // 08h, 09h, 10h, ..., 22h
    days: Day[]
    values: {
      // segunda-08h00: busy, segunda-09h00: available, ...
      [key: string]: TimeboxValue
    }
  }
}


const Schedule: React.FC<ScheduleProps> = ({ data }) => {
  const { days, times } = data
  const [values, setValues] = useState(data.values)

  const setTimeBoxValue = (dayTime: string) => {
    // chamar função que atualiza valor do timebox globalmente    
    const oldValue = values[dayTime]
    let newValue: TimeboxValue = undefined
    if (oldValue === 'available')
      newValue = 'busy'
    if (oldValue === 'busy')
      newValue = undefined
    if (oldValue === undefined)
      newValue = 'available'

    setValues({ ...values, [`${dayTime}`]: newValue })
  }

  return (
    <>
      <h1>Schedule</h1>

      <Container cols={days.length + 1} rows={times.length}>
        <div id="top-left-corner">Horário \ Dia </div>

        {days.map(day => (
          // se a tela for muito pequena, mostrar apenas a primeira letra
          <div className="day" key={`day-time-${day}`}> {day[0]} </div>
        ))}

        {times.map(time => {
          return (
            [<div className="time" key={time}> {time} </div>]
              .concat(days.map(day => (
                <Timebox
                  onClick={() => setTimeBoxValue(`${day}-${time}`)}
                  key={`${day}-${time}`}
                  value={values[`${day}-${time}`]}
                />
              ))))
        })}
      </Container>
    </>
  )
}

export default Schedule