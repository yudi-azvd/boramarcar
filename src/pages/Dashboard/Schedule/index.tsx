import Day from "@/types/Day"
import DayTime from "@/types/DayTime"
import Time from "@/types/Time"
import TimeboxValue from "@/types/TimeboxValue"
import { useSocket } from "@/hooks/socket"
import { notification } from "antd"
import React, { useState } from "react"
import { Container, Timebox } from "./style"


export interface ScheduleProps {
  visible: boolean
  data: {
    times: Time[],
    days: Day[]
    values: {
      [key in DayTime]: TimeboxValue
    }
  }
}


const Schedule: React.FC<ScheduleProps> = ({ data, visible }) => {
  const { days, times } = data
  const { socket } = useSocket()
  const [step, setStep] = useState(1)
  const [values, setValues] = useState(data.values)

  const setTimeBoxValue = (dayTime: DayTime) => {
    // chamar função que atualiza valor do timebox globalmente    
    const oldValue = values[dayTime]
    let newValue: TimeboxValue = undefined
    if (oldValue === 'available')
      newValue = 'busy'
    if (oldValue === 'busy')
      newValue = undefined
    if (oldValue === undefined)
      newValue = 'available'

    const [day, time] = dayTime.split('-') as [Day, Time]
    socket.emit('change-timebox-value', { day, time, newValue })
    setValues({ ...values, [dayTime]: newValue })
  }

  const onMouseDown = (dayTime: DayTime) => {
    notification.error({
      message: dayTime
    })
  }
  
  const onMouseUp = (dayTime: DayTime) => {
    notification.error({
      message: dayTime
    })
  }

  const onTouchStart = (dayTime: DayTime) => {
    notification.success({
      message: dayTime,
      description: 'touch START'
    })
  }

  const onTouchEnd = (dayTime: DayTime) => {
    notification.success({
      message: dayTime,
      description: 'touch END'
    })
  }

  return (
    <>
      <Container cols={days.length + 1} rows={times.length + 1} visible={visible} >
        <div id="top-left">Horário \ Dia </div>

        {days.map(day => (
          // se a tela for muito pequena, mostrar apenas a primeira letra
          <div className="day" key={`day-${day}`}> {day[0]} </div>
        ))}

        {times.map(time => {
          return (
            [<div className="time" key={time}> {time} </div>]
              .concat(days.map(day => {
                const dayTime = `${day}-${time}` as DayTime
                return (
                  <Timebox
                    onMouseDown={() => onMouseDown(dayTime)}
                    onMouseUp={() => onMouseUp(dayTime)}
                    onTouchStart={() => onTouchStart(dayTime)}
                    onTouchEnd={() => onTouchEnd(dayTime)}
                    onClick={() => setTimeBoxValue(dayTime)}
                    onDoubleClick={() => console.log('double click')}
                    key={dayTime}
                    value={values[dayTime]}
                  />
                )
              })))
        })}
      </Container>
    </>
  )
}

export default Schedule