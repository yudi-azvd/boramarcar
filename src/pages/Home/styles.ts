import styled from "styled-components";

export const Container = styled.div`
  height: 100vh;
  display: flex;
  align-items: stretch;
`

export const Content = styled.div`
  display: flex;
  flex-direction: column;
  margin-top: 60px;
  align-items: center;
  width: 100%;

  > img, > h1 {
    margin-bottom: 116px;
  }

  div {
    display: flex;
    flex-direction: column;

    button + button {
      margin-top: 30px;
    }
  }
`