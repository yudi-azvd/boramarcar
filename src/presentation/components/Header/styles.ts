import styled from 'styled-components'

export const Container = styled.header`
  width: 90%;
  margin: 0 auto;
  margin-top: 16px;
  margin-bottom: 32px;

  display: flex;
  justify-content: space-between;
  align-items: center;

  div.auth-info {
    padding: 0 12px;
  }
`
