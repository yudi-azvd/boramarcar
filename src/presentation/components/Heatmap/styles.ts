import { TimeBoxValue } from '@/types'
import styled from 'styled-components'

export const timeboxColors: {
  [key in 'available' | 'busy']: string
} = {
  // undefined: #DDDDDD
  available: '#18DC86',
  busy: '#E95F63',
}

interface TimeboxProps {
  value?: TimeBoxValue
}

export const Timebox = styled.div<TimeboxProps>`
  background: ${props => props.value ? timeboxColors[props.value] : '#DDDDDD'};
  cursor: pointer;
  
  & {
    transition: all 0.2s ease;
  }

  &:hover {
    -webkit-filter: brightness(90%);
    filter: brightness(90%);
  }
`

interface ContainerProps {
  cols: number
  rows: number
  visible: boolean
}

export const Container = styled.div<ContainerProps>`
  width: 100%;
  height: 90vh;
  display: grid;
  grid-template-columns: repeat(${props => props.cols}, 1fr);
  grid-template-rows: repeat(${props => props.rows}, 1fr);
  column-gap: 4px;
  row-gap: 4px;

  > div {
    user-select: none;
    -moz-user-select: none;
    -khtml-user-select: none;
    -webkit-user-select: none;
    -o-user-select: none;
    border-radius: 4px;
  }

  > div#top-left {
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    border-right: 1px solid rgba(0, 0, 0, 0.1);
  }

  > div.day {
    position: relative;

    &::after {
      position: absolute;
      content: '';
      background: rgba(0, 0, 0, 0.1);
      height: 100%;
      width: 1px;
      right: -3px;
    }
  }

  > div.time {
    position: relative;

    &::after {
      position: absolute;
      content: '';
      background: rgba(0, 0, 0, 0.1);
      height: 1px;
      width: 100%;
      bottom: -3px;
    }
  }

  > div#top-left, > div.time, > div.day {
    border-radius: 5px;
  }

  > div.day, div.time {
    display: flex;
    justify-content: center;
    align-items: center;
  }

  ${Timebox} {
    opacity: ${props => props.visible ? 1 : 0};
  }
`