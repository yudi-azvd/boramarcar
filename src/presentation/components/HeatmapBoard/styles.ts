import styled from 'styled-components'

interface HeatmapTimeboxProps {
  color: string
}

export const HeatmapTimebox = styled.div<HeatmapTimeboxProps>`
  background: ${({ color }) => color};
  border: 1px solid rgba(0, 0, 0, 0.1);
  
  & {
    transition: all 0.2s ease;
  }
`

interface ContainerProps {
  cols: number
  rows: number
  visible: boolean
}

export const Container = styled.div<ContainerProps>`
  width: 100%;
  height: 90vh;
  display: grid;
  grid-template-columns: repeat(${props => props.cols}, 1fr);
  grid-template-rows: repeat(${props => props.rows}, 1fr);
  column-gap: 4px;
  row-gap: 4px;

  > div {
    user-select: none;
    -moz-user-select: none;
    -khtml-user-select: none;
    -webkit-user-select: none;
    -o-user-select: none;
    border-radius: 4px;
  }

  > div.day, div.time {
    display: flex;
    justify-content: center;
    align-items: center;
  }

  ${HeatmapTimebox} {
    opacity: ${props => props.visible ? 1 : 0};
  }
`