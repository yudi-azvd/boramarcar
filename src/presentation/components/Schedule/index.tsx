import { ScheduleRepository } from "@/contracts"
import { Day, DayTime, Time, TimeBoxValue } from "@/types"
import { useEffect, useState } from "react"
import { Container, Timebox } from "./styles"

interface ScheduleProps {
  times: Time[]
  days: Day[]
  userId: string
  scheduleRepository: ScheduleRepository
}

const dict: {
  [key in Day]: string
} = {
  'Sunday': 'Domingo',
  'Monday': 'Segunda',
  'Tuesday': 'Terça',
  'Wednesday': 'Quarta',
  'Thursday': 'Quinta',
  'Friday': 'Sexta',
  'Saturday': 'Sábado'
}

const Schedule: React.FC<ScheduleProps> = ({ days, times, userId, scheduleRepository }) => {
  const [values, setValues] = useState({} as {
    [key in DayTime]?: TimeBoxValue
  })

  async function setTimeBoxValue(dayTime: DayTime): Promise<void> {
    const oldValue = values[dayTime]
    let newValue: TimeBoxValue = undefined

    if (oldValue === 'available')
      newValue = 'busy'
    if (oldValue === 'busy')
      newValue = undefined
    if (oldValue === undefined)
      newValue = 'available'

    // Se não tiver await o teste falha. Parece que o testes de Schedule.spec.tsx 
    // intererem uns nos outros.
    await scheduleRepository.update({ userId, dayTime, timeboxValue: newValue })
    setValues({ ...values, [dayTime]: newValue })
  }

  useEffect(() => {
    async function getAll() {
      const timeboxes = await scheduleRepository.getAll({ userId })
      setValues(timeboxes)
    }

    getAll()
  }, [])

  return (
    <Container cols={days.length + 1} rows={times.length + 1} visible>
      <div id="top-left" />

      {days.map(day => (
        <div className="day" key={`sch-${day}`}> {dict[day][0]} </div>
      ))}

      {times.map(time => (
        [<div className="time" key={`sch-${time}`} > {time} </div>].concat(
          days.map(day => {
            const dayTime: DayTime = `${day}-${time}`
            return (
              <Timebox
                onClick={() => setTimeBoxValue(dayTime)}
                // Precisa dessa classe para os testes. Ou escolhe outro seletor 
                // mais fácil lá em Schedule.spec/makeSut 
                className="timebox"
                id={`sch-${dayTime}`}
                key={`sch-${dayTime}`}
                value={values[dayTime] ?? undefined}
              />
            )
          })
        )
      ))}
    </Container>
  )
}

export default Schedule