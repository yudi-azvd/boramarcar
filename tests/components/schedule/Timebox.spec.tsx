import { render } from '@testing-library/react'
import '@testing-library/jest-dom'
import 'jest-styled-components'

import { Timebox } from '@/presentation/components/Schedule/styles'

interface SutProps {
  value?: 'available' | 'busy'
}

interface Sut {
  sut: HTMLDivElement
}

function makeSut({ value }: SutProps): Sut {
  const { container } = render(<Timebox id="timebox-id" value={value} />)
  return { sut: container.querySelector('#timebox-id') as HTMLDivElement }
}

describe('Timebox', () => {
  it('should be grey when its value is undefined', () => {
    const { sut } = makeSut({ value: undefined })

    expect(sut).toMatchSnapshot()
    expect(sut).toHaveStyle(`background: #DDDDDD`)
  })

  it('should be green when its value is available', () => {
    const { sut } = makeSut({ value: 'available' })
    
    expect(sut).toMatchSnapshot()
    expect(sut).toHaveStyle(`background: #18DC86`)
  })
  
  it('should be red when its value is busy', () => {
    const { sut } = makeSut({ value: 'busy' })

    expect(sut).toMatchSnapshot()
    expect(sut).toHaveStyle(`background: #E95F63`)
  })
})
