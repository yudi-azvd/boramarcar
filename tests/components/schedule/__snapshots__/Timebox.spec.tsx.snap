// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timebox should be green when its value is available 1`] = `
.c0 {
  background: #18DC86;
}

<div
  class="c0"
  id="timebox-id"
  value="available"
/>
`;

exports[`Timebox should be red when its value is busy 1`] = `
.c0 {
  background: #E95F63;
}

<div
  class="c0"
  id="timebox-id"
  value="busy"
/>
`;

exports[`Timebox should be transparent when its value is notdefined 1`] = `
.c0 {
  background: transparent;
}

<div
  class="c0"
  id="timebox-id"
  value="notdefined"
/>
`;
